platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1505,45 +1505,63 @@ private void assertUsageIsMutable() {
-     * </ul>
+     * violate these conditions.  However, it may emit a warning on redundant changes if a special flag is set.
      * <p>
      * The eventual goal is that all configuration usage be specified upon creation and immutable
      * thereafter.
      */
-    private void maybePreventChangingUsage(String methodName, boolean current, boolean newValue) {
+    private void checkChangingUsage(String methodName, boolean current, boolean newValue) {
         if (hasAllUsages()) {
             // We currently allow configurations with all usages -- those that are created with
             // `create` and `register` -- to have mutable roles. This is likely to change in the future
             // when we deprecate any configuration with mutable roles.
             return;
         }
 
+        boolean redundantChange = current == newValue;
+
         // Error will be thrown later. Don't emit a duplicate warning.
-        if (!usageCanBeMutated && (current != newValue)) {
+        if (!usageCanBeMutated && !redundantChange) {
             return;
         }
 
         // KGP continues to set the already-set value for a given usage even though it is already set
-        boolean redundantChange = current == newValue;
-
-        // KGP disables `consumable` on detached configurations even though this is not necessary
-        boolean disableUsageForDetached = isDetachedConfiguration() && !newValue;
-
-        // This property exists to allow KGP to test whether they have properly resolved this deprecation.
-        // This property WILL be removed without warning.
-        if ((redundantChange || disableUsageForDetached) &&
-            !Boolean.getBoolean("org.gradle.internal.deprecation.preliminary.Configuration.redundantUsageChangeWarning.enabled")
-        ) {
-            return;
+        // This property exists to allow KGP to test whether they have properly stopped making unnecessary redundant
+        // changes to detachedConfigurations.
+        // This property WILL be removed without warning and should be removed in Gradle 9.x.
+        boolean extraWarningsEnabled = Boolean.getBoolean("org.gradle.internal.deprecation.preliminary.Configuration.redundantUsageChangeWarning.enabled");
+
+        if (redundantChange) {
+            // Remove this condition in Gradle 9.x and warn on every redundant change, in Gradle 10.0 this should fail.
+            if (extraWarningsEnabled) {
+                warnAboutChangingUsage(methodName, newValue);
+            }
+        } else {
+            if (isDetachedConfiguration()) {
+                // This is an actual change, and permitting it is not desired behavior, but we haven't deprecated
+                // changing detached confs as of 9.0, so we have to permit even non-redundant changes to them,
+                // but we can at least warn if the flag is set.
+                // Remove this check and warn on every actual change to a detached conf in Gradle 9.x, in Gradle 10.0 this should fail.
```

Comment:
This is not completely true. 
The old logic is `isDetachedConfiguration() && !newValue` where the new condition here is only `isDetachedConfiguration()`.

So we only permitted setting properties of the detached configuration to false, but deprecated setting the properties of detached configurations to true. 

This PR changes the previous condition to permitting _all_ changes to a detached configuration. We maintain the old logic so it continues to be an error to set any usage of a detached configuration to true.